test/freenet/client/filter/ContentFilterTest.java

```diff
@@ -27,360 +32,363 @@
+        assertEquals(META_CHARSET_LOWER_RES, htmlFilter(META_CHARSET_LOWER));
+    }
+
+    @Test
+    public void testEvilCharset() throws IOException {
+        // This is why we need to disallow characters before <html> !!
+        String s = "<html><body><a href=\"http://www.google.com/\">Blah</a>";
+        String end = "</body></html>";
+        String alt = "<html><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-16\"></head><body><a href=\"http://www.freenetproject.org/\">Blah</a></body></html>";
+        if ((s.length() + end.length()) % 2 == 1) {
+            s += " ";
+        }
+        s = s + end;
+        byte[] buf = s.getBytes(StandardCharsets.UTF_8);
+        byte[] utf16bom = new byte[]{(byte) 0xFE, (byte) 0xFF};
+        byte[] bufUTF16 = alt.getBytes(StandardCharsets.UTF_16);
+        byte[] total = new byte[buf.length + utf16bom.length + bufUTF16.length];
+        System.arraycopy(utf16bom, 0, total, 0, utf16bom.length);
+        System.arraycopy(buf, 0, total, utf16bom.length, buf.length);
+        System.arraycopy(bufUTF16, 0, total, utf16bom.length + buf.length, bufUTF16.length);
+        HTMLFilter filter = new HTMLFilter();
+        boolean failed = false;
+        List<RuntimeException> failures = new ArrayList<>();
+        try (
+            FileOutputStream fos = new FileOutputStream("output.utf16")
+        ){
+            ArrayBucket out = new ArrayBucket();
+            filter.readFilter(new ArrayBucket(total).getInputStream(), out.getOutputStream(), "UTF-16", null, null, null);
+            fos.write(out.toByteArray());
+            failed = true;
+            failures.add(new RuntimeException("Filter accepted dangerous UTF8 text with BOM as UTF16! (HTMLFilter)"));
+        } catch (DataFilterException e) {
+            System.out.println("Failure: " + e);
+            e.printStackTrace();
+            if (e.getCause() != null) {
+                e.getCause().printStackTrace();
+            }
+            // Ok.
+        }
+        try (
+            FileOutputStream fos = new FileOutputStream("output.filtered")
+        ){
+            ArrayBucket out = new ArrayBucket();
+            FilterStatus fo = ContentFilter.filter(new ArrayBucket(total).getInputStream(), out.getOutputStream(), "text/html", null, null, null);
+            fos.write(out.toByteArray());
+            failed = true;
+            failures.add(new RuntimeException("Filter accepted dangerous UTF8 text with BOM as UTF16! (ContentFilter) - Detected charset: " + fo.charset));
+        } catch (DataFilterException e) {
+            System.out.println("Failure: " + e);
+            e.printStackTrace();
+            if (e.getCause() != null) {
+                e.getCause().printStackTrace();
+            }
+            // Ok.
+        }
+
+        if (failed) {
+            try (FileOutputStream fos = new FileOutputStream("unfiltered")) {
+                fos.write(total);
+            }
+            throw failures.get(0);
```

Comment:
Gosh, at some point we need to take a fine-toothed comb and unsuck all of these tests…